File: SimulateMoments.jl
using Distributions, LinearAlgebra, Optim, FastGaussQuadrature, Plots, Interpolations, Random, Statistics, DataFrames, GLM, Printf, CSV

include("ModelFunctions.jl")
include("EstimationFunctions.jl")

# Parameter Bounds
ϵ_bounds   = (4.0, 20.0)
σν2_bounds = (0.01, 0.3)
δ_bounds   = (0.01, 0.2)
μω_bounds  = (0.01, 0.2)
γ_bounds   = (0.8, 1.5)
ση2_bounds = (0.1, 1.0)
ρ_bounds   = (0.00, 0.9)

n_param_points = 250
param_bounds = [
    γ_bounds,
    μω_bounds,
    ση2_bounds,
    ρ_bounds,
    σν2_bounds,
    ϵ_bounds,
    δ_bounds
]

param_vectors = halton_param_vectors(param_bounds, n_param_points; seed=212311)

# Halton draws index 2 as μω = μη/(1-ρ); recover μη = μω * (1-ρ).
# Index 2 = μω, index 4 = ρ, index 7 = log(δ) (order: γ, μω→μη, ση2, ρ, σν2, ϵ, δ).
for v in param_vectors
    v[2] = log(v[2]) * (1.0 - v[4])
    # v[7] = exp(v[7])
end

# Slice param_vectors by SLURM array task ID (falls back to full set if not in a job array)
task_id    = parse(Int, get(ENV, "SLURM_ARRAY_TASK_ID",    "1"))
n_tasks    = parse(Int, get(ENV, "SLURM_ARRAY_TASK_COUNT", "1"))
chunk_size = ceil(Int, length(param_vectors) / n_tasks)
i_start    = (task_id - 1) * chunk_size + 1
i_end      = min(task_id * chunk_size, length(param_vectors))
param_vectors = param_vectors[i_start:i_end]

output_path = n_tasks > 1 ?
    "SimulatedData/moments_$(lpad(task_id, 3, '0')).csv" :
    "SimulatedData/moments.csv"

println("Task $(task_id)/$(n_tasks): grid points $(i_start)–$(i_end) ($(length(param_vectors)) total)")
# output_path = "../SimulatedData/moments.csv"

df_out = compute_moments_on_grid(
    param_vectors;
    n_firms=500,
    n_years=20,
    seed=212311,
    max_value_iterations=500,
    grid_size = 200, scale = 1.0, size = 100.0, solve_tol = 1e-2,
    output_path=output_path
)

n_ok = sum(.!df_out.failed)
println("Sweep complete. $(n_ok) / $(nrow(df_out)) points succeeded.")
println("Saved: $(output_path)")

# # Drop grid points with implausible inventory-to-sales ratios
# isr_threshold = 20.0
# n_before = nrow(df_out)
# df_out = df_out[coalesce.(df_out.avg_isr, Inf) .<= isr_threshold, :]
# n_dropped = n_before - nrow(df_out)
# n_dropped > 0 && @printf("Dropped %d / %d grid points with avg_isr > %.1f\n",
#                           n_dropped, n_before, isr_threshold)

# Summary output
fail_fraction = sum(df_out.failed) / nrow(df_out)
@printf("Failure fraction: %.4f (%d / %d)\n", fail_fraction, sum(df_out.failed), nrow(df_out))


df_success = df_out[(.!df_out.failed ).& (df_out.avg_isr.<4.0), :]
if nrow(df_success) == 0
    println("No successful simulations; moment summaries unavailable.")
else
    println("\nMoment summaries (successful simulations, $(nrow(df_success))):")
    println("moment, min,p10, p25, median, p75, p90,max")

    moment_cols = [:avg_isr, :var_log1p_isr, :avg_gross_margin, :γ_OLS, :ρ_ω, :σ_η2, :avg_opex_sales]
    for col in moment_cols
        vals = collect(skipmissing(df_success[!, col]))
        vals = vals[isfinite.(vals)]
        if isempty(vals)
            @printf("%s, NaN, NaN, NaN, NaN, NaN, NaN, NaN\n", String(col))
        else
            @printf("%s,%.6f, %.6f, %.6f, %.6f, %.6f,%.6f, %.6f\n",
                    String(col),
                    minimum(vals),
                    quantile(vals, 0.10),
                    quantile(vals, 0.25),
                    quantile(vals, 0.50),
                    quantile(vals, 0.75),
                    quantile(vals, 0.90),
                    maximum(vals))
        end
    end
end

# df_out[!,:μω] = exp.(df_out[!,:μη]./(1 .- df_out[!,:ρω]))

# df_success = df_out[.!df_out.failed, :]
# # df_success = df_success[df_success[!,:γ_OLS].>0,:]
# if nrow(df_success) == 0
#     println("No successful simulations; moment summaries unavailable.")
# else
#     println("\nParameter summaries (successful simulations only):")
#     println("parameter, min,p10, p25, median, p75, p90,max")

#     moment_cols = [:γ, :μω, :ση2, :ρω, :σν2, :ϵ, :δ]
#     for col in moment_cols
#         vals = collect(skipmissing(df_success[!, col]))
#         vals = vals[isfinite.(vals)]
#         if isempty(vals)
#             @printf("%s, NaN, NaN, NaN, NaN, NaN, NaN, NaN\n", String(col))
#         else
#             @printf("%s,%.6f, %.6f, %.6f, %.6f, %.6f,%.6f, %.6f\n",
#                     String(col),
#                     minimum(vals),
#                     quantile(vals, 0.10),
#                     quantile(vals, 0.25),
#                     quantile(vals, 0.50),
#                     quantile(vals, 0.75),
#                     quantile(vals, 0.90),
#                     maximum(vals))
#         end
#     end
# end

File: PythonEstimation/simulatemoments.py
"""simulatemoments.py

Python translation of SimulateMoments.jl.

Generates a Halton parameter grid, slices it by job-array task ID, and
evaluates all 7 estimation moments sequentially over the slice.  Inner
parallelism (Numba threads) handles each solve_value_function call.
Outer parallelism is via the job array.

Supports PBS Pro and SLURM job arrays.
  PBS  : reads PBS_ARRAY_INDEX (task id) and PBS_ARRAY_N_TASKS (total tasks,
         set explicitly in the PBS script via -v PBS_ARRAY_N_TASKS=N).
  SLURM: reads SLURM_ARRAY_TASK_ID and SLURM_ARRAY_TASK_COUNT.
  None : runs as a single task over all parameter vectors.

Usage (single node):
    python simulatemoments.py

Usage (PBS Pro job array):
    qsub -J 1-N -v PBS_ARRAY_N_TASKS=N simulatemoments.pbs

Usage (SLURM job array):
    #SBATCH --array=1-N
    python simulatemoments.py
"""

import os
import math

import numpy as np

from simulation_functions import halton_param_vectors, compute_moments_on_grid

# ---------------------------------------------------------------------------
# Parameter bounds  (must match SimulateMoments.jl)
# ---------------------------------------------------------------------------
# Order: [γ, μω→μη, ση2, ρ, σν2, ε, δ]
# μω is drawn then converted: μη = log(μω) * (1 − ρ)
BOUNDS = [
    (4.0,  30.0),   # ε
    (0.01,  0.4),   # σν2
    (0.005, 0.2),   # δ
    (0.01,  0.2),   # μω  (converted to μη below)
    (0.8,   1.5),   # γ
    (0.1,   1.0),   # ση2
    (0.00,  0.9),   # ρ
]

# Matching the field order expected by compute_moments_on_grid:
# [γ, μη, ση2, ρω, σν2, ε, δ]
# We draw in BOUNDS order then rearrange after the μω→μη transform.

N_PARAM_POINTS = 2500
SEED           = 212311

N_FIRMS   = 500
N_YEARS   = 20
BURN_IN   = 100
GRID_SIZE = 200
SCALE     = 1.0
SIZE      = 100.0
SOLVE_TOL = 1e-2
MAX_VFI   = 500

OUT_DIR ="SimulatedData"

# ---------------------------------------------------------------------------
# Generate Halton draws in BOUNDS order
# ---------------------------------------------------------------------------
raw_vectors = halton_param_vectors(BOUNDS, N_PARAM_POINTS, seed=SEED)

# BOUNDS order: [ε, σν2, δ, μω, γ, ση2, ρ]
# Apply μω → μη = log(μω) * (1 − ρ)  (index 3 = μω, index 6 = ρ)
param_vectors = []
for v in raw_vectors:
    eps_i   = v[0]
    snu2_i  = v[1]
    delta_i = v[2]
    mu_omega_i = v[3]
    gamma_i = v[4]
    sig2_i  = v[5]
    rho_i   = v[6]
    mu_eta_i = math.log(mu_omega_i) * (1.0 - rho_i)
    # Target order for compute_moments_on_grid: [γ, μη, ση2, ρω, σν2, ε, δ]
    param_vectors.append(np.array([gamma_i, mu_eta_i, sig2_i, rho_i, snu2_i, eps_i, delta_i]))

# ---------------------------------------------------------------------------
# Job-array slicing  (PBS Pro and SLURM)
# ---------------------------------------------------------------------------
if "PBS_ARRAY_INDEX" in os.environ:
    task_id = int(os.environ["PBS_ARRAY_INDEX"])
    n_tasks = int(os.environ.get("PBS_ARRAY_N_TASKS", "1"))
elif "SLURM_ARRAY_TASK_ID" in os.environ:
    task_id = int(os.environ["SLURM_ARRAY_TASK_ID"])
    n_tasks = int(os.environ["SLURM_ARRAY_TASK_COUNT"])
else:
    task_id = 1
    n_tasks = 1

chunk_size = math.ceil(len(param_vectors) / n_tasks)
i_start    = (task_id - 1) * chunk_size          # 0-based
i_end      = min(task_id * chunk_size, len(param_vectors))
param_slice = param_vectors[i_start:i_end]

if n_tasks > 1:
    output_path = os.path.join(OUT_DIR, f"moments_{task_id:03d}.csv")
else:
    output_path = os.path.join(OUT_DIR, "moments.csv")

print(
    f"Task {task_id}/{n_tasks}: grid points {i_start+1}–{i_end} "
    f"({len(param_slice)} total)"
)

# ---------------------------------------------------------------------------
# Grid sweep
# ---------------------------------------------------------------------------
df_out = compute_moments_on_grid(
    param_slice,
    n_firms=N_FIRMS,
    n_years=N_YEARS,
    seed=SEED,
    output_path=output_path,
    max_value_iterations=MAX_VFI,
    grid_size=GRID_SIZE,
    scale=SCALE,
    size=SIZE,
    solve_tol=SOLVE_TOL,
)

# ---------------------------------------------------------------------------
# Summary
# ---------------------------------------------------------------------------
n_ok = int((~df_out["failed"]).sum())
print(f"Sweep complete. {n_ok} / {len(df_out)} points succeeded.")
print(f"Saved: {output_path}")

fail_frac = df_out["failed"].mean()
print(f"Failure fraction: {fail_frac:.4f} ({df_out['failed'].sum()} / {len(df_out)})")

df_success = df_out[(~df_out["failed"]) & (df_out["avg_isr"] < 4.0)]
if len(df_success) == 0:
    print("No successful simulations; moment summaries unavailable.")
else:
    print(f"\nMoment summaries (successful simulations, n={len(df_success)}):")
    print(f"{'moment':<22} {'min':>10} {'p10':>10} {'p25':>10} {'median':>10} {'p75':>10} {'p90':>10} {'max':>10}")
    moment_cols = [
        "avg_isr", "var_log1p_isr", "avg_gross_margin",
        "gamma_OLS", "rho_omega_ar1", "sigma_eta2_ar1", "avg_opex_sales",
    ]
    for col in moment_cols:
        vals = df_success[col].dropna()
        vals = vals[np.isfinite(vals)]
        if len(vals) == 0:
            print(f"  {col:<20}  all NaN")
            continue
        qs = np.quantile(vals, [0.10, 0.25, 0.50, 0.75, 0.90])
        print(
            f"  {col:<20}  "
            f"{vals.min():10.6f} {qs[0]:10.6f} {qs[1]:10.6f} "
            f"{qs[2]:10.6f} {qs[3]:10.6f} {qs[4]:10.6f} {vals.max():10.6f}"
        )
